src/Features/Core/Portable/NavigateTo/NavigateToSearcher.cs

```diff
@@ -81,111 +104,187 @@ internal async Task SearchAsync()
-                var priorityDocs = ImmutableArray.Create(activeDocument).AddRange(visibleDocsFromProject);
+            // We didn't have any items reported *and* we weren't fully loaded.  If it turns out
+            // that some of our projects were using cached data then we can try searching them
+            // again, but this tell them to use the latest data.  The ensures the user at least gets
+            // some result instead of nothing.
+            var projectsUsingCache = projectResults.Where(t => t.location == NavigateToSearchLocation.Cache).SelectAsArray(t => t.project);
+            await ProcessProjectsAsync(ImmutableArray.Create(projectsUsingCache), isFullyLoaded: true, cancellationToken).ConfigureAwait(false);
 
-                // Search the active project first.  That way we can deliver results that are
-                // closer in scope to the user quicker without forcing them to do something like
-                // NavToInCurrentDoc
-                await Task.Run(() => SearchAsync(activeProject, priorityDocs, seenItems, isFullyLoaded), _cancellationToken).ConfigureAwait(false);
+            // we did a full uncached search.  return true to indicate that no message should be
+            // shown to the user.
+            return true;
+        }
+
+        /// <summary>
+        /// Returns a sequence of groups of projects to process.  The sequence is in priority order,
+        /// and all projects in a particular group should be processed before the next group.  This
+        /// allows us to associate CPU resources in likely areas the user wants, while also still
+        /// allowing for good parallelization.
+        /// </summary>
+        private ImmutableArray<ImmutableArray<Project>> GetOrderedProjectsToProcess()
+        {
+            // If we're only searching the current doc, we don't need to examine anything else but that.
+            if (_searchCurrentDocument)
+            {
+                var project = _currentDocument?.Project;
+                return project == null
+                    ? ImmutableArray<ImmutableArray<Project>>.Empty
+                    : ImmutableArray.Create(ImmutableArray.Create(project));
+            }
+
+            using var result = TemporaryArray<ImmutableArray<Project>>.Empty;
+
+            using var _ = PooledHashSet<Project>.GetInstance(out var processedProjects);
+
+            // First, if there's an active document, search that project first, prioritizing that
+            // active document and all visible documents from it.
+            if (_activeDocument != null)
+            {
+                processedProjects.Add(_activeDocument.Project);
+                result.Add(ImmutableArray.Create(_activeDocument.Project));
+            }
+
+            // Next process all visible docs that were not from the active project.
+            using var buffer = TemporaryArray<Project>.Empty;
```

Comment:
seems like this could be slightly simpler if the active document was just the first visible document, would just need this group

Reply:
taken offline.